### PR TITLE
Implement ask_for_reset for HTMLSelectElement.

### LINKED ISSUE #7774: Implement 'ask for a reset' algorithm for <select> elements
https://html.spec.whatwg.org/multipage/#ask-for-a-reset

### PR DESCRIPTION
**Pull request by dagnir**
Fixes #7774

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/7963)

### CODE REVIEW

**Comment by frewsxcv** [comment]
`ask_for_reset` (the function you just created) should get called [here](https://github.com/servo/servo/blob/babeed998d297d2d7badc06ba9c8f90e42239b4f/components/script/dom/htmloptionelement.rs#L147-L148)

**Comment by dagnir** [comment]
Hmmm `ask_for_reset` calls `HTMLOptionElement::SetSeledcted` so that will make them mutually recursive.  Is this intentional?  Or should I be setting `HTMLOptionElement::selectedness` directly somehow?


---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by frewsxcv** [comment]
> Or should I be setting HTMLOptionElement::selectedness directly somehow?

You could add a `set_selectedness` method on `HTMLOptionElement` so that you could set it directly


---







Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmloptionelement.rs, line 153 \[r4\]](https://reviewable.io:443/reviews/servo/servo/7963#-K0JWjJEt3G-9f_4Jt7A)** ([raw file](https://github.com/servo/servo/blob/4599d424a6b4fc8665b83d635dc36dfee6a4fd2d/components/script/dom/htmloptionelement.rs#L153)):</sup>
Instead of doing `.find(|a| a.is_htmlselectelement)` and then `HTMLSelectElementCast::to_root(..).unwrap`, use `filter_map` with `HTMLSelectElementCast::to_root`

http://doc.servo.org/std/iter/trait.Iterator.html#method.filter_map

Let me know if you need help 


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by dagnir** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

<sup>**[components/script/dom/htmloptionelement.rs, line 153 \[r4\]](https://reviewable.io:443/reviews/servo/servo/7963#-K0JWjJEt3G-9f_4Jt7A)** ([raw file](https://github.com/servo/servo/blob/4599d424a6b4fc8665b83d635dc36dfee6a4fd2d/components/script/dom/htmloptionelement.rs#L153)):</sup>
Thanks, changed to `filter_map`.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmlselectelement.rs, line 78 \[r5\]](https://reviewable.io:443/reviews/servo/servo/7963#-K0JoQIP2vSwpPy3Fr_I)** ([raw file](https://github.com/servo/servo/blob/3f529b0025bd01c4fa82defbd147dab8d9cb0a68/components/script/dom/htmlselectelement.rs#L78)):</sup>
This could be simplified to `is_none` and `get_enabled_state`


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmlselectelement.rs, line 72 \[r5\]](https://reviewable.io:443/reviews/servo/servo/7963#-K0JrEWd6-U2Hfn1thln)** ([raw file](https://github.com/servo/servo/blob/3f529b0025bd01c4fa82defbd147dab8d9cb0a68/components/script/dom/htmlselectelement.rs#L72)):</sup>
Since we implement [`Deref` on `Root`](http://doc.servo.org/servo/script/dom/bindings/js/struct.Root.html), you [don't need to explicitly](https://doc.rust-lang.org/nightly/book/deref-coercions.html#deref-and-method-calls) call `.r()`. It's the same reason one can do `String::from("hello").len()` even though `len()` is a `&str` method (`String` `Deref`s to `&str`)


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmlselectelement.rs, line 78 \[r5\]](https://reviewable.io:443/reviews/servo/servo/7963#-K0JoQIP2vSwpPy3Fr_I)** ([raw file](https://github.com/servo/servo/blob/3f529b0025bd01c4fa82defbd147dab8d9cb0a68/components/script/dom/htmlselectelement.rs#L78)):</sup>
Actually, `get_enabled_state` might be something different 


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by dagnir** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[components/script/dom/htmlselectelement.rs, line 72 \[r5\]](https://reviewable.io:443/reviews/servo/servo/7963#-K0JrEWd6-U2Hfn1thln)** ([raw file](https://github.com/servo/servo/blob/3f529b0025bd01c4fa82defbd147dab8d9cb0a68/components/script/dom/htmlselectelement.rs#L72)):</sup>
Removed unneeded calls to `r()`.


---

<sup>**[components/script/dom/htmlselectelement.rs, line 78 \[r5\]](https://reviewable.io:443/reviews/servo/servo/7963#-K0JoQIP2vSwpPy3Fr_I)** ([raw file](https://github.com/servo/servo/blob/3f529b0025bd01c4fa82defbd147dab8d9cb0a68/components/script/dom/htmlselectelement.rs#L78)):</sup>
Simplified check.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by frewsxcv** [comment]
FYI, the latest commit doesn't pass the style checker, one of the lines is too long

https://travis-ci.org/servo/servo/jobs/84851573

**Comment by dagnir** [comment]
Hi @frewsxcv, is there anything else needed for this PR?

**Comment by frewsxcv** [comment]
Have you tried running the web-platform-tests (`test-wpt`) test suite to see if any tests unexpectedly pass or fail. If you need help doing that, check out [this link](https://github.com/servo/servo/tree/master/tests/wpt#running-the-tests)). If after running the tests, nothing changes, a regression test will need to be written to make sure algorithm this doesn't break in the future. If you need help with any of this or want clarification, please let me know, I'll be happy to help! I'm currently doing IRL non-Servo work at the moment, but when I get home I'll actually read through the spec and make sure your implementation looks correct.

**Comment by dagnir** [comment]
I'm running `test-wpt` now.  If you would show me how to get started with writing the regression test for this, I'd appreciate it!

Thanks!

**Comment by frewsxcv** [comment]
(to anyone who is reading this PR and wondering why we haven't responded, we just chatted 1 on 1 on IRC)

**Comment by dagnir** [comment]
@frewsxcv, I've added sow tests, please take a look.  Only the third one passes at the moment because `ask_for_reset` doesn't get called on node insert and remove.

**Comment by eefriedman** [comment]
The relevant hooks for node insert and remove are `bind_to_tree()` and `unbind_from_tree()`.

**Comment by dagnir** [comment]
Thanks!

**Comment by frewsxcv** [comment]
FYI, in case you missed it, there are merge conflicts that will eventually need to be addressed

**Comment by dagnir** [comment]
Branch is now up-to-date.

**Comment by dagnir** [comment]
Hmm it seems `unbind_from_tree` is called after the node is already removed, so I can't iterate over the ancestors to find the `select` (from `HTMLOptionElement`).  Is there a "will remove" or similar hook I can use?

Also, looks like `html/semantics/forms/the-select-element/select-remove.html` has multiple failures atm, so maybe there's more work involved for removal to work?

**Comment by jdm** [comment]
I wonder if we should be including the previous parent node as an argument or something. That wouldn't work for notifying children of the root unbound node, though. Do you have thoughts on this, @Ms2ger?

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmlselectelement.rs, line 76 \[r6\]](https://reviewable.io:443/reviews/servo/servo/7963#-K17OtqnqAvpbb-Z9FtC)** ([raw file](https://github.com/servo/servo/blob/fb5f6c3169de9ace787b511889cf28f73cb0f2c0/components/script/dom/htmlselectelement.rs#L76)):</sup>
The spec doesn't say anything about 'display size' == 0, right?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmlselectelement.rs, line 75 \[r6\]](https://reviewable.io:443/reviews/servo/servo/7963#-K17PPRaBIr01w5O-ABK)** ([raw file](https://github.com/servo/servo/blob/fb5f6c3169de9ace787b511889cf28f73cb0f2c0/components/script/dom/htmlselectelement.rs#L75)):</sup>
Instead of keeping track of the number of elements that are 'selected' (which is a number we don't really care about and might as well be a boolean), why not just check here if `last_selected.is_none()`


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by eefriedman** [comment]
> The spec doesn't say anything about 'display size' == 0, right?

The display size of a select is always greater than zero.  That's the only reasonable way to interpret https://html.spec.whatwg.org/multipage/forms.html#concept-select-size.

**Comment by dagnir** [comment]
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmlselectelement.rs, line 76 \[r6\]](https://reviewable.io:443/reviews/servo/servo/7963#-K17OtqnqAvpbb-Z9FtC)** ([raw file](https://github.com/servo/servo/blob/fb5f6c3169de9ace787b511889cf28f73cb0f2c0/components/script/dom/htmlselectelement.rs#L76)):</sup>
Yeah I changed it to include 0 because `Size()` defaults 0: https://github.com/servo/servo/blob/941c06eb99533c6c1d55b4ad3b51804badf48a23/components/script/dom/htmlselectelement.rs#L28.  I checked and Gecko also defaults to 0, even if one option is being displayed.  Don't know if this is a bug or not 


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by dagnir** [comment]
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmlselectelement.rs, line 76 \[r6\]](https://reviewable.io:443/reviews/servo/servo/7963#-K17OtqnqAvpbb-Z9FtC)** ([raw file](https://github.com/servo/servo/blob/fb5f6c3169de9ace787b511889cf28f73cb0f2c0/components/script/dom/htmlselectelement.rs#L76)):</sup>
Actually I think this is fine if `Size() == 0` means it's absent...spec says

>  if the size attribute is absent, then the element's display size is 4 if the element's multiple content attribute is present, and 1 otherwise.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmlselectelement.rs, line 76 \[r6\]](https://reviewable.io:443/reviews/servo/servo/7963#-K17OtqnqAvpbb-Z9FtC)** ([raw file](https://github.com/servo/servo/blob/fb5f6c3169de9ace787b511889cf28f73cb0f2c0/components/script/dom/htmlselectelement.rs#L76)):</sup>
So there are two possibilities here I see:

* `size` attribute is explicitly specified
  * From the spec:  

> "The size attribute, if specified, must have a value that is a valid non-negative integer **greater than zero**."

* `size` attribute is not specified
  * From the spec

> "If applying those rules to the attribute's value is not successful, or if the size attribute is absent, then the element's display size is 4 if the element's multiple content attribute is present, and **1 otherwise**"

I don't see a scenario where it's possible for it to be zero *from the spec* (ignore our codebase and Gecko codebase)


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmlselectelement.rs, line 76 \[r6\]](https://reviewable.io:443/reviews/servo/servo/7963#-K17OtqnqAvpbb-Z9FtC)** ([raw file](https://github.com/servo/servo/blob/fb5f6c3169de9ace787b511889cf28f73cb0f2c0/components/script/dom/htmlselectelement.rs#L76)):</sup>
Also, something to keep in mind that I forgot to clarify in my previous comment here. `size` and `display size` are two separate concepts. It appears to be possible for `size` to be `0`, but not `display size`. Considering it's a separate concept, I think it's appropriate to have a separate method that returns the `display size` (with a link to the `display size` portion of the spec adjacent to it). That method just returns the `size` attribute if it's not `0`, those default values otherwise


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by SimonSapin** [comment]
> "The size attribute, if specified, must have a value that is a valid non-negative integer greater than zero."

I don’t have any context, but this sounds like an author requirements. Sooner or later, a some website will violate that requirement. However wrong this is, a browser has to handle it somehow. Another part of the spec should define how, through implementation requirements.

**Comment by frewsxcv** [comment]
@dagnir Let me know if you need any assistance with this or if you need any clarification on the next steps

**Comment by dagnir** [comment]
Thanks @frewsxcv.  I haven't had a change to work on this the past couple days, but I should have an updated patch up tonight or tomorrow.  I'll ping you on IRC if I have questions

**Comment by dagnir** [comment]
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/dom/htmlselectelement.rs, line 76 \[r6\]](https://reviewable.io:443/reviews/servo/servo/7963#-K17OtqnqAvpbb-Z9FtC)** ([raw file](https://github.com/servo/servo/blob/fb5f6c3169de9ace787b511889cf28f73cb0f2c0/components/script/dom/htmlselectelement.rs#L76)):</sup>
Done.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmlselectelement.rs, line 108 \[r11\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1MK3rwehboMZm1Kq_a)** ([raw file](https://github.com/servo/servo/blob/fcfc3918a4e53931ef8cd61a1f6a51d9bfc2e20f/components/script/dom/htmlselectelement.rs#L108)):</sup>
The way you wrote this is fine, but you can also do:

```
match self.Size() {
  0 if self.Multiple() => 4,
  0 => 1,
  s => s,
}
```


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmlselectelement.rs, line 82 \[r11\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1MJVz-efchWQhrLrCG)** ([raw file](https://github.com/servo/servo/blob/fcfc3918a4e53931ef8cd61a1f6a51d9bfc2e20f/components/script/dom/htmlselectelement.rs#L82)):</sup>
Using `unwrap` is generally a bad practice. A couple easy ways to avoid this:

```
if Some(last_slected) = last_selected {
  ...
} else {
  ...
}
```

or even better:

```
match last_selected {
  Some(last_selected) => ...,
  None => ...,
}
```


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by frewsxcv** [comment]
<sup>**[tests/wpt/web-platform-tests/html/semantics/forms/the-select-element/select-ask-for-reset.html, line 39 \[r9\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1MmoOcVqJp6lSMMHn_)** ([raw file](https://github.com/servo/servo/blob/663801ed79eefc6d20511cade9efb6cdf6e905dc/tests/wpt/web-platform-tests/html/semantics/forms/the-select-element/select-ask-for-reset.html#L39)):</sup>
Nit: extra indentation here


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by frewsxcv** [comment]
Looks good to me after my latest comments! Thanks for all your hard work :)

**Comment by dagnir** [comment]
Thanks for all the help!.  Also FYI, I marked the first test as `expected: FAIL` since as discussed earlier in this thread, I can't get to the `select` in `unbind_from_tree` to call `ask_for_reset`.


---

Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/dom/htmlselectelement.rs, line 82 \[r11\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1MJVz-efchWQhrLrCG)** ([raw file](https://github.com/servo/servo/blob/fcfc3918a4e53931ef8cd61a1f6a51d9bfc2e20f/components/script/dom/htmlselectelement.rs#L82)):</sup>
Done.


---

<sup>**[tests/wpt/web-platform-tests/html/semantics/forms/the-select-element/select-ask-for-reset.html, line 39 \[r9\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1MmoOcVqJp6lSMMHn_)** ([raw file](https://github.com/servo/servo/blob/663801ed79eefc6d20511cade9efb6cdf6e905dc/tests/wpt/web-platform-tests/html/semantics/forms/the-select-element/select-ask-for-reset.html#L39)):</sup>
Done.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

File: tests/wpt/metadata/html/semantics/forms/the-select-element/select-ask-for-reset.html.ini [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[select-ask-for-reset.html]
```
Comment (frewsxcv) [inline tests/wpt/metadata/html/semantics/forms/the-select-element/select-ask-for-reset.html.ini]: Is there any reason why this test is marked as failing?

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmloptionelement.rs, line 145 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1U7ZHePn1tEG8neqE4)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmloptionelement.rs#L145)):</sup>
Where in the spec does it talk about picking options upon setting `selected`?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmlselectelement.rs, line 85 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1U96ZKuqAf82o1BLWn)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L85)):</sup>
Does this method correspond to a particular section in the spec?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

File: tests/wpt/metadata/html/semantics/forms/the-select-element/select-ask-for-reset.html.ini (hunk 1 above)
Comment (jdm) [inline tests/wpt/metadata/html/semantics/forms/the-select-element/select-ask-for-reset.html.ini]: See first paragraph of previous comment.

**Comment by dagnir** [comment]
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmlselectelement.rs, line 85 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1U96ZKuqAf82o1BLWn)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L85)):</sup>
Yes, it's here https://html.spec.whatwg.org/multipage/forms.html#concept-select-pick (and the paragraph after it).  I'm not sure now whether I should have left it to be fixed in another issue, but the second test doesn't pass without it.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by jdm** [comment]
@dagnir With regards to the unbind_from_tree issue, you might be interested in the changes in https://github.com/servo/servo/compare/nox:finish-ranges .

**Comment by dagnir** [comment]
@jdm awesome, I'll keep an eye out to see when it lands.

**Comment by eefriedman** [comment]
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmloptionelement.rs, line 147 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f6MLR0aExVK2S5Crt)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmloptionelement.rs#L147)):</sup>
Weird indentation.


---

<sup>**[components/script/dom/htmloptionelement.rs, line 208 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f62_g3NGUR0srr6TT)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmloptionelement.rs#L208)):</sup>
This is missing a call to the ask-for-reset algorithm.  (Sort of an edge case... but consider what happens if you insert an option which isn't selected into an empty select.)


---

<sup>**[components/script/dom/htmlselectelement.rs, line 67 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f33GEuYH5nQAt2vVK)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L67)):</sup>
Whitespace.


---

<sup>**[components/script/dom/htmlselectelement.rs, line 73 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f3QWOC3L1VnCL-ip3)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L73)):</sup>
`if let Some(last_selected) = `...


---

<sup>**[components/script/dom/htmlselectelement.rs, line 85 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1U96ZKuqAf82o1BLWn)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L85)):</sup>
It would be nice to include that URL as a comment before the definition of this function.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by dagnir** [comment]
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmloptionelement.rs, line 147 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f6MLR0aExVK2S5Crt)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmloptionelement.rs#L147)):</sup>
Done.


---

<sup>**[components/script/dom/htmloptionelement.rs, line 208 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f62_g3NGUR0srr6TT)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmloptionelement.rs#L208)):</sup>
Done.


---

<sup>**[components/script/dom/htmlselectelement.rs, line 67 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f33GEuYH5nQAt2vVK)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L67)):</sup>
Done.


---

<sup>**[components/script/dom/htmlselectelement.rs, line 73 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f3QWOC3L1VnCL-ip3)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L73)):</sup>
From what I understood from @frewsxcv's comment here: https://reviewable.io/reviews/servo/servo/7963#-K1MJVz-efchWQhrLrCG, this was preferable, unless I misunderstood 


---

<sup>**[components/script/dom/htmlselectelement.rs, line 85 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1U96ZKuqAf82o1BLWn)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L85)):</sup>
Done.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by eefriedman** [comment]
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/dom/htmlselectelement.rs, line 73 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f3QWOC3L1VnCL-ip3)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L73)):</sup>
`if let` is generally preferred over `is_none()`+`unwrap()`; I don't think that review comment was specifically suggesting the opposite.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by dagnir** [comment]
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/dom/htmlselectelement.rs, line 73 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f3QWOC3L1VnCL-ip3)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L73)):</sup>
Hmm is it possible you saw an outdated revision?  This is what it currently looks like https://github.com/dagnir/servo/commit/b5e991c89e11be2467b30d1212052c57e21c81d6


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by eefriedman** [comment]
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/dom/htmlselectelement.rs, line 73 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f3QWOC3L1VnCL-ip3)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L73)):</sup>
Oh, right, sorry, I somehow got flipped to the wrong revision just now.  Still, `if let` is generally preferred over match.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmlselectelement.rs, line 73 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f3QWOC3L1VnCL-ip3)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L73)):</sup>
"Oh, right, sorry, I somehow got flipped to the wrong revision just now. Still, if let is generally preferred over match."

I disagree, I think they're both perfectly valid. `match` gives you the benefit of erroring when all cases are not covered.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by frewsxcv** [comment]
Thanks for looking at this @eefriedman, I've been busy these past few days and haven't had time to review this.

**Comment by eefriedman** [comment]
> I disagree, I think they're both perfectly valid. match gives you the benefit of erroring when all cases are not covered.

You're right.  I'll narrow it a bit; "if let" is preferred over "match" for `Option<T>`.

**Comment by dagnir** [comment]
@frewsxcv @eefriedman okay to leave, or should I change it?

**Comment by eefriedman** [comment]
Please change it.  Also, please refactor the identical code in `SetSelected` and `ask_for_reset`.

**Comment by dagnir** [comment]
@eefriedman I'm not sure what identical code you're referring to; unless you mean the option element traversal in `ask_for_reset` and `pick_option`?

**Comment by eefriedman** [comment]
SetSelected:

```
+        if let Some(select) = self.upcast::<Node>().ancestors()
+                .filter_map(Root::downcast::<HTMLSelectElement>).next() {
+            if selected {
+                select.pick_option(self);
+            }
+            select.ask_for_reset();
+        }
```

ask_for_reset:

```
+        if let Some(select) = self.upcast::<Node>().ancestors()
+                .filter_map(Root::downcast::<HTMLSelectElement>)
+                .next() {
+            if self.Selected() {
+                select.pick_option(self);
+            }
+            select.ask_for_reset();
+        }
```

**Comment by eefriedman** [comment]
Err, oops, I meant SetSelected and bind_to_tree.

**Comment by dagnir** [comment]
Thanks, done.


---

Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)

**Comment by frewsxcv** [comment]
@bors-servo retry

* https://github.com/servo/servo/issues/8236

**Comment by Ms2ger** [comment]
For future reference:

> "if let" is preferred over "match" for `Option<T>`

is not true. It is preferred only for (some) cases where only a single variant has code associated with it.

**Pull request merged**